### Empirical Risk Minimisation

The goal of ERM is to minimise the average loss on the training data:

$$
\hat{\theta} = \arg\min_{\theta}
\frac{1}{n}\sum_{i=1}^{n}
L(y_i, f_{\theta}(x_i))
$$

where:

- $n$ = number of training examples
- $y_i$ = actual value
- $f_{\theta}(x_i)$ = model prediction
- $L$ = loss function
- $\theta$ = model parameters

# Simple implementation from scratch using Linear Regression and Mean Squared Error

In [2]:
import numpy as np
import matplotlib.pyplot as plt

# Training data
X = np.array([1, 2, 3, 4, 5])
y = np.array([2, 4, 5, 8, 10])

# Model: y_hat = w*x + b
def predict(X, w, b):
    return w * X + b

# Mean Squared Error
def loss(X, y, w, b):
    y_pred = predict(X, w, b)
    return np.mean((y - y_pred) ** 2)

# Search for parameters that minimize empirical risk
best_w = None
best_b = None
min_loss = float("inf")

for w in np.arange(0, 4, 0.01):
    for b in np.arange(-2, 3, 0.01):
        
        current_loss = loss(X, y, w, b)
        
        if current_loss < min_loss:
            min_loss = current_loss
            best_w = w
            best_b = b

        #This part is called ERM where we search for the parameters w and b that produce the lowest average training loss.

print("Best w:", best_w)
print("Best b:", best_b)
print("Minimum empirical risk:", min_loss)

# Predictions
y_pred = predict(X, best_w, best_b)

print("Predictions:", y_pred)


Best w: 2.0
Best b: -0.1999999999999984
Minimum empirical risk: 0.1600000000000001
Predictions: [1.8 3.8 5.8 7.8 9.8]


Exactly. Those results show that your Empirical Risk Minimization (ERM) code has found the values of 
w
 and 
b
 that minimize the training loss.

### Empirical Risk Minimization

The linear model is:

$$
\hat{y} = wx + b
$$

ERM minimizes the empirical risk:

$$
\hat{\theta} =
\arg\min_{\theta}
\frac{1}{n}
\sum_{i=1}^{n}
L(y_i, f_{\theta}(x_i))
$$

For our model, the optimal parameters are:

$$
w = 2.0,\qquad b = -0.2
$$

Therefore:

$$
\hat{y} = 2x - 0.2
$$

The minimum empirical risk is:

$$
R_{\text{emp}} = 0.16
$$